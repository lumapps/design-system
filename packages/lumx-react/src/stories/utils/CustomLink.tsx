/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
