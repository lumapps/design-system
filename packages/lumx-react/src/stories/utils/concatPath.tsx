export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
