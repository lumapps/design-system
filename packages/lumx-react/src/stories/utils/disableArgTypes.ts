export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
