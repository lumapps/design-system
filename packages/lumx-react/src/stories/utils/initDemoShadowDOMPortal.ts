/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
