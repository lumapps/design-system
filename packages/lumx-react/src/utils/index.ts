/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
