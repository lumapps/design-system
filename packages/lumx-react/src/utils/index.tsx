export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
