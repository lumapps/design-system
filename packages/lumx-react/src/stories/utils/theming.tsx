import React from 'react';
import { Predicate } from '@lumx/react/utils/type';
import { FlexBox } from '@lumx/react';

type Property = string | { value: string };
interface PropertyTree {
    [key: string]: PropertyTree | Property;
}
type Entry = [key: string, value: string];
type Variables = Record<string, string>;

function asValue(prop?: PropertyTree | Property): string | undefined {
    if (typeof prop === 'string') {
        return prop;
    }
    if (typeof prop === 'object' && 'value' in prop) {
        return prop.value as string;
    }
    return undefined;
}

/** Convert property tree into a list of args entries ([key, value]) */
function* toArgEntries(props: PropertyTree | Property, parent = ''): Iterable<Entry> {
    const value = asValue(props);
    if (value) {
        yield [parent, value];
    } else {
        for (const [key, prop] of Object.entries(props)) {
            const newParent = parent ? `${parent}-${key}` : key;
            yield* toArgEntries(prop, newParent);
        }
    }
}
function* parseVariableEntries(cssString: string, strict = false) {
    for (const line of cssString.split(/(?<=;)/)) {
        if (!line.trim()) continue;
        const match = line.match(/^\s*(--lumx[^:]+)\s*:\s*([^;]+);/m);
        if (!match) {
            if (strict) throw new Error(`Could not parse \`${line}\``);
            else continue;
        }
        const [, variableName, variableValue] = match;
        yield [variableName, variableValue];
    }
}

function VariableEditor({
    variable: [key, value],
    setVariable,
    defaultStyle,
}: {
    variable: Entry;
    setVariable: (newVariable: Entry) => void;
    defaultStyle: Variables;
}) {
    const onChange = React.useCallback<React.ChangeEventHandler<HTMLInputElement>>(
        (evt) => {
            setVariable([evt.target.name, evt.target.value]);
        },
        [setVariable],
    );

    return (
        <FlexBox as="label" orientation="horizontal" hAlign="center" style={{ whiteSpace: 'nowrap' }}>
            <span style={{ flexShrink: 0 }}>{key}</span>:
            <input
                name={key}
                onChange={onChange}
                style={{ border: 0, padding: 8, fieldSizing: 'content', display: 'inline' } as any}
                value={defaultStyle[key] === value ? '' : value}
                placeholder={value}
            />
            ;
        </FlexBox>
    );
}

function VariablesEditor({
    style,
    defaultStyle,
    setStyle,
}: {
    style: Variables;
    defaultStyle: Variables;
    setStyle: (fn: (newStyle: Variables) => Variables) => void;
}) {
    const setVariable = React.useCallback(
        ([key, value]: Entry) => {
            setStyle((prevStyle) => ({ ...prevStyle, [key]: value }));
        },
        [setStyle],
    );
    return (
        <details style={{ fontFamily: 'monospace' }} className="lumx-spacing-margin-vertical-huge">
            <summary className="lumx-typography-subtitle2">Edit variables</summary>
            {Object.entries(style).map((variable) => (
                <VariableEditor
                    key={variable[0]}
                    defaultStyle={defaultStyle}
                    variable={variable}
                    setVariable={setVariable}
                />
            ))}
        </details>
    );
}

const FILTER_PROPS = ['theme', 'emphasis', 'size'];

export function withTheming<P extends PropertyTree>({
    properties,
    values,
}: {
    properties?: P;
    values?: string | Partial<P>;
}) {
    const initialStyle = values
        ? Object.fromEntries(
              typeof values === 'string'
                  ? parseVariableEntries(values)
                  : toArgEntries(values as PropertyTree, '--lumx'),
          )
        : [];
    const defaultStyle = Object.fromEntries(properties ? Array.from(toArgEntries(properties, '--lumx')) : []);

    const varNames = new Set(Object.keys(initialStyle).concat(Object.keys(defaultStyle)));

    function filteredStyles(args: any) {
        // Create predicate
        const filterArgs = Object.entries(args)
            .map(([argName, argValue]) => {
                if (typeof argValue === 'string' && FILTER_PROPS.includes(argName)) {
                    return (varName: string) =>
                        // Has the given arg with value
                        varName.match(new RegExp(`${argName}-${argValue}`)) ||
                        (argName === 'emphasis' && argValue === 'medium' && varName.match(/emphasis-selected/)) ||
                        // Or does not have the arg at all
                        !varName.match(new RegExp(`${argName}-\\w+`));
                }
                return false;
            })
            .filter(Boolean) as Array<Predicate<string>>;
        const styles = Object.fromEntries(
            Array.from(varNames.values())
                .map((varName) => {
                    if (!filterArgs.every((filter) => filter(varName))) return null as any;
                    return [varName, initialStyle[varName] || defaultStyle[varName]];
                })
                .filter(Boolean),
        );
        return () => styles;
    }

    return (Story: any, ctx: any) => {
        const [style, setStyle] = React.useState<Variables>(filteredStyles(ctx.args));

        return (
            <>
                <Story args={{ ...ctx.args, style: { ...ctx.args.style, ...style } }} />
                <VariablesEditor style={style} defaultStyle={defaultStyle} setStyle={setStyle} />
            </>
        );
    };
}
