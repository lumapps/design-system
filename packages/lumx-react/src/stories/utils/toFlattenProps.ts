import castArray from 'lodash/castArray';
import { concatPath } from './concatPath';

type Props = Record<string, any>;
type OneOrMoreProps = Props | Array<Props>;

/**
 * Build a flat props object from the given nested props
 *
 * @example toFlattenProps({ foo: { bar: 4 } }) // => { 'foo.bar': 4 }
 * @example toFlattenProps({ foo: [{ bar: 4 }, { bar: 5 }] }) // => { 'foo[0].bar': 4, 'foo[0].bar': 5 }
 */
export function toFlattenProps(props: { [prefix: string]: OneOrMoreProps }): Props {
    const out: Props = {};
    for (const [prefix, oneOrMoreProps] of Object.entries(props)) {
        const propsArray = castArray(oneOrMoreProps);

        for (let i = 0; i < propsArray.length; i += 1) {
            const path = Array.isArray(oneOrMoreProps) ? concatPath(prefix, i) : prefix;
            const subProps = propsArray[i];

            for (const [key, value] of Object.entries(subProps)) {
                out[concatPath(path, key)] = value;
            }
        }
    }
    return out;
}
