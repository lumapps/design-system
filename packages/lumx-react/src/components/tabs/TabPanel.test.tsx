import { CommonSetup, Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';
import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import React, { ReactElement } from 'react';

import { CLASSNAME, TabPanel, TabPanelProps } from './TabPanel';
import { setupTabProviderMocks } from './test.mocks';

// Mock useTabProviderContext.
jest.mock('./state', () => ({ useTabProviderContext: jest.fn() }));

/**
 * Define the overriding properties waited by the `setup` function.
 */
type SetupProps = Partial<TabPanelProps>;

/**
 * Defines what the `setup` function will return.
 */
interface Setup extends CommonSetup {
    props: SetupProps;

    /**
     * The <div> element that wraps tab.
     */
    wrapper: Wrapper;
}

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props  The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup = ({ ...propsOverrides }: SetupProps = {}, shallowRendering: boolean = true): Setup => {
    const props: TabPanelProps = { ...propsOverrides, children: 'Tab panel content' };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<TabPanel {...props} />);

    return { props, wrapper };
};

describe(`<${TabPanel.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            setupTabProviderMocks({ index: 1 });
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });

        it('should render isActive state', () => {
            setupTabProviderMocks({ index: 2, isActive: true });
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });

        it('should not render children when not active and lazy', () => {
            setupTabProviderMocks({ isLazy: true, isActive: false });
            const { wrapper } = setup();
            expect(wrapper.find('div').props().children).toBeFalsy();
        });

        it('should render children when active and lazy', () => {
            setupTabProviderMocks({ isLazy: true, isActive: true });
            const { wrapper } = setup();
            expect(wrapper.find('div').props().children).toBeTruthy();
        });

        it('should render children when not active and not lazy', () => {
            setupTabProviderMocks({ isActive: false, isLazy: false });
            const { wrapper } = setup();
            expect(wrapper.find('div').props().children).toBeTruthy();
        });
    });

    // Common tests suite.
    commonTestsSuite(
        (...args: any[]) => {
            setupTabProviderMocks();
            return setup(...args);
        },
        { className: 'wrapper', prop: 'wrapper' },
        { className: CLASSNAME },
    );
});
