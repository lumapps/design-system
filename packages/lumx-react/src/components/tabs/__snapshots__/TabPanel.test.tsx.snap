// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabPanel> Snapshots and structure should render correctly 1`] = `
<div
  aria-labelledby="tab-1"
  className="lumx-tab-panel"
  id="tab-panel-1"
  role="tabpanel"
  tabIndex={0}
>
  Tab panel content
</div>
`;

exports[`<TabPanel> Snapshots and structure should render isActive state 1`] = `
<div
  aria-labelledby="tab-2"
  className="lumx-tab-panel lumx-tab-panel--is-active"
  id="tab-panel-2"
  role="tabpanel"
  tabIndex={0}
>
  Tab panel content
</div>
`;
